src/librustc/mir/interpret/queries.rs

```diff
@@ -2,24 +2,46 @@ use super::{ConstEvalResult, ErrorHandled, GlobalId};
 
 use crate::mir;
 use crate::ty::subst::{InternalSubsts, SubstsRef};
-use crate::ty::{self, TyCtxt};
+use crate::ty::{self, Ty, TyCtxt};
 use rustc_hir::def_id::DefId;
 use rustc_span::Span;
 
 impl<'tcx> TyCtxt<'tcx> {
     /// Evaluates a constant without providing any substitutions. This is useful to evaluate consts
     /// that can't take any generic arguments like statics, const items or enum discriminants. If a
-    /// generic parameter is used within the constant `ErrorHandled::ToGeneric` will be returned.
+    /// generic parameter is used within the constant `ErrorHandled::TooGeneric` will be returned.
     pub fn const_eval_poly(self, def_id: DefId) -> ConstEvalResult<'tcx> {
-        // In some situations def_id will have substitutions within scope, but they aren't allowed
-        // to be used. So we can't use `Instance::mono`, instead we feed unresolved substitutions
-        // into `const_eval` which will return `ErrorHandled::ToGeneric` if any of them are
-        // encountered.
-        let substs = InternalSubsts::identity_for_item(self, def_id);
-        let instance = ty::Instance::new(def_id, substs);
-        let cid = GlobalId { instance, promoted: None };
-        let param_env = self.param_env(def_id).with_reveal_all();
-        self.const_eval_global_id(param_env, cid, None)
+        let ty = self.type_of(def_id);
+        debug!("const_eval_poly: running on {:?}: {:?}", def_id, ty);
+
+        // FIXME: This check could be smarter / faster.
+        let is_generic = |t: Ty<'_>| -> bool {
```

Comment:
this isn't the only way that it can be too generic, the final decision on whether it is too generic is made during const eval. why was it necessary to add this check? just for speed?

Reply:
It's to address [this test case](https://github.com/rust-lang/rust/blob/master/src/test/ui/consts/const-eval/ice-generic-assoc-const.rs). In this case, the call to `layout_of` fails
https://github.com/rust-lang/rust/blob/5da2e53f471f5eea60f534845b90c97ecabb71cb/src/librustc_mir/interpret/place.rs#L310

because `T` isn't known yet -- we're not running on a specific instance of `T`. The error is bubbled up, [transformed into a validation failure](https://github.com/rust-lang/rust/blob/e0f5df017368dc3f7cb458fc6d5a5e0420e1d2e5/src/librustc_mir/interpret/validity.rs#L471-L475), and is caught in the catch-all here:

https://github.com/rust-lang/rust/blob/e0f5df017368dc3f7cb458fc6d5a5e0420e1d2e5/src/librustc_mir/const_eval/eval_queries.rs#L213-L218

As I understand it, the test case failure boils down to not having a concrete type to substitute into `T` -- this is something we currently don't run into, because  `check_binary_op` currently runs when `const_prop` reaches a use of the value, so there's always some actual type to substitute in to get a layout for.

Other TooGeneric failures are currently caught in that catch-all error emit above. My initial thought when I ran into this was to modify the functions along the callstack so instead of catching the error, turning it into a validation failure and reporting the error, they can throw `TooGeneric` all the way up and let the caller decide what to do with them (so, in the case of check_const, it would just ignore the TooGeneric and move on). This seemed a bit too specific, though, and checking here seemed sufficient -- and yes, maybe even a little bit faster.

I've just noticed the `struct_error` here is actually const eval's `struct_error`, which looks like could already bubble TooGeneric up. Hmm -- I'll have a look at how it goes if the layout_of caller does a check to throw TooGeneric instead of undefined pointer.

Reply:
yea, I think we should fix this in const eval. although the last time I tried, I did it wrongly. maybe it suffices to use a `UserFacing` `ParamEnv` for associated constants?

Reply:
Just having this change didn't end up solving the issue -- as it turns out, it wasn't const eval that was failing in the first place, but validation afterwards. I resolved this by changing validation to not fail on pointers to Params (which should hopefully only be reached when checking unused constants).